packages/rest/src/rest.server.ts

```diff
@@ -338,20 +343,42 @@ export class RestServer extends Context implements Server, HttpServerLike {
   }
 
   protected _setupHandlerIfNeeded() {
-    // TODO(bajtos) support hot-reloading of controllers
-    // after the app started. The idea is to rebuild the HttpHandler
-    // instance whenever a controller was added/deleted.
-    // See https://github.com/strongloop/loopback-next/issues/433
     if (this._httpHandler) return;
+    if (this._controllersAndRoutesView == null) {
+      // Watch for binding events
+      this._controllersAndRoutesView = this.createView(
+        binding =>
+          filterByKey(RestBindings.API_SPEC.key)(binding) ||
+          (filterByKey(/^(controllers|routes)\..+/)(binding) &&
+            // Exclude controller routes to avoid circular events
+            !filterByTag(RestTags.CONTROLLER_ROUTE)(binding)),
+      );
+    }
+    this._createHttpHandler();
+
+    // See https://github.com/strongloop/loopback-next/issues/433
+    this._controllersAndRoutesView.on('refresh', () => {
```

Comment:
From what I remember, ContextView is rather heavy-weight and may involve async operations. Have you considered using ContextListener instead? 

I am also not sure about the behavior of the `refresh` event, is it fired on the same tick of the event loop when the bindings were added/removed? I am concerned that if `ContextView` was delaying those events because to the next tick of event loop or Promise microqueue, then it would create a race condition where a request may arrive after we have registered the new controller but before the changes were picked up by the routing layer.

Reply:
`ContextView` is notified to refresh after the binding is fully populated with `ctx.bind(...).tag(...)`. It will be executed in another tick.

When a new controller is added by `app.controller()`, the following steps will happen:

1. A new binding for the controller is added (sync)
2. A context `bind` event is emitted (sync) and a `notification` event is added to a queue
3. In another tick, the notification is processed to invoke ContextView's `observe()`, which calls `refresh()` and emits `refresh` event.

I think the race condition is acceptable as `app.controller` call does not expect the controller will be available in the same tick.

Reply:
I believe your work on `ContextListener` included a way how to be notified about changes in the binding metadata?

Setting timing issues aside, I feel `ContextView` solution is too heavy weight. IIRC, `ContextView` is maintaining a list of all matched bindings, which unnecessarily increases memory consumption, considering that we are not using that list at all.

Anyhow, as long as ContextView vs. ContextListener is just an implementation detail that can be changed later, I am fine with landing the current approach.

Reply:
I switched to `ContextObserver`.